src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs

```diff
@@ -114,84 +101,86 @@ public bool ContentSet
         /// <summary>
         /// Generates the <see cref="TagHelperOutput"/>'s start tag.
         /// </summary>
-        /// <returns><c>string.Empty</c> if <see cref="TagName"/> is <c>null</c> or whitespace. Otherwise, the
+        /// <returns><c>Null</c> if <see cref="TagName"/> is <c>null</c> or whitespace. Otherwise, the
         /// <see cref="string"/> representation of the <see cref="TagHelperOutput"/>'s start tag.</returns>
         public string GenerateStartTag()
         {
             // Only render a start tag if the tag name is not whitespace
             if (_isTagNameNullOrWhitespace)
             {
-                return string.Empty;
+                return null;
             }
 
-            var sb = new StringBuilder();
-
-            sb.Append('<')
-              .Append(TagName);
+            var startTag = new StringBuilder();
+            startTag.Append("<");
+            startTag.Append(TagName);
 
             foreach (var attribute in Attributes)
             {
                 var value = WebUtility.HtmlEncode(attribute.Value);
-                sb.Append(' ')
-                  .Append(attribute.Key)
-                  .Append("=\"")
-                  .Append(value)
-                  .Append('"');
+                startTag.Append(" ");
+                startTag.Append(attribute.Key);
+                startTag.Append("=\"");
+                startTag.Append(value);
+                startTag.Append("\"");
             }
 
             if (SelfClosing)
             {
-                sb.Append(" /");
+                startTag.Append(" /");
             }
 
-            sb.Append('>');
+            startTag.Append(">");
 
-            return sb.ToString();
+            return startTag.ToString();
         }
 
         /// <summary>
         /// Generates the <see cref="TagHelperOutput"/>'s <see cref="PreContent"/>.
         /// </summary>
-        /// <returns><c>string.Empty</c> if <see cref="TagName"/> is not <c>null</c> or whitespace
-        /// and <see cref="SelfClosing"/> is <c>true</c>. Otherwise, <see cref="PreContent"/>.</returns>
-        public string GeneratePreContent()
+        /// <returns><c>Null</c> if <see cref="TagName"/> is not <c>null</c> or whitespace
+        /// and <see cref="SelfClosing"/> is <c>true</c>.
+        /// Otherwise, returns <see cref="ITextWriterCopyable"/>.</returns>
+        public ITextWriterCopyable GeneratePreContent()
         {
             if (!_isTagNameNullOrWhitespace && SelfClosing)
             {
-                return string.Empty;
+                return null;
             }
 
-            return PreContent;
+            return (ITextWriterCopyable)PreContent;
```

Comment:
IMO good reason to separate into a field of `DefaultTagHelperContent`. It's more obvious that the code is correct, but still exposes the API we want.